Recover from missing param list in function definitions

clubby789 opened the pull request:
Addresses the other issue mentioned in #108109

clubby789 committed:
Recover from missing param list in function definitions

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6671822851/job/18134525436) [(plain)](https://github.com/rust-lang/rust/commit/de98267eb87d43c8b54cfb98b7f5e6c2548a4f53/checks/18134525436/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=clubby789
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_fae99a4b-ae71-43cf-9505-51df360ded94
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=c8ac302933a8a8304c60bd4c06270cca48f8597c
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_fae99a4b-ae71-43cf-9505-51df360ded94
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_fae99a4b-ae71-43cf-9505-51df360ded94
GITHUB_TRIGGERING_ACTOR=clubby789
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/117298/merge
GITHUB_WORKFLOW_SHA=c8ac302933a8a8304c60bd4c06270cca48f8597c
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---

9 LL |     pub struct Foo { text }
10    |         ++++++
11 
- error: expected one of `(` or `<`, found `{`
+   --> $DIR/recovered-block.rs:17:8
+    |
+    |
+ LL |     Foo { text: "".to_string() }
+ 
+ error: struct literal body without path
13   --> $DIR/recovered-block.rs:17:9
14    |
14    |
15 LL |     Foo { text: "".to_string() }

-    |         ^ expected one of `(` or `<`
+    |
+    |
+ help: you might have forgotten to add the struct literal inside the block
+    |
+ LL |     Foo { SomeStruct { text: "".to_string() } }
17 
- error: aborting due to 2 previous errors
+ error[E0308]: mismatched types
+   --> $DIR/recovered-block.rs:15:17
+   --> $DIR/recovered-block.rs:15:17
+    |
+ LL | pub fn bar() -> Foo {
+    |        ---      ^^^ expected `Foo`, found `()`
+    |        |
+    |        implicitly returns `()` as its body has no tail or `return` expression
+ error: aborting due to 4 previous errors
+ 
+ For more information about this error, try `rustc --explain E0308`.
20 
---
   |
LL |     pub Foo { text }
   |        ^
   |
help: add `struct` here to parse `Foo` as a public struct
LL |     pub struct Foo { text }
   |         ++++++

error: missing parameters for function definition
error: missing parameters for function definition
##[error]  --> /checkout/tests/ui/mismatched_types/recovered-block.rs:17:8
   |
LL |     Foo { text: "".to_string() }

error: struct literal body without path
##[error]  --> /checkout/tests/ui/mismatched_types/recovered-block.rs:17:9
   |
   |
LL |     Foo { text: "".to_string() }
   |
   |
help: you might have forgotten to add the struct literal inside the block
   |
LL |     Foo { SomeStruct { text: "".to_string() } }

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/mismatched_types/recovered-block.rs:15:17
   |
   |
LL | pub fn bar() -> Foo {
   |        ---      ^^^ expected `Foo`, found `()`
   |        |
   |        implicitly returns `()` as its body has no tail or `return` expression
error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0308`.
------------------------------------------
------------------------------------------


---- [ui] tests/ui/parser/issues/issue-105366.rs stdout ----
diff of stderr:

- error: you might have meant to write `impl` instead of `fn`
+ error: missing parameters for function definition
+   --> $DIR/issue-105366.rs:5:13
3    |
4 LL | fn From<i32> for Foo {
4 LL | fn From<i32> for Foo {
-    | ^^
+    |             ^ help: add a parameter list
+ 
+ error: expected one of `->`, `where`, or `{`, found keyword `for`
6    |
6    |
- help: replace `fn` with `impl` here
- LL | impl From<i32> for Foo {
-    | ~~~~
+ LL | fn From<i32> for Foo {
+ LL | fn From<i32> for Foo {
+    |              ^^^ expected one of `->`, `where`, or `{`
- error: aborting due to previous error
+ error: aborting due to 2 previous errors
13 
14 
---

2 
3 struct Foo;
4 
- impl From<i32> for Foo {
+ fn From<i32>() for Foo {
6     //~^ ERROR you might have meant to write `impl` instead of `fn`
7     fn from(_a: i32) -> Self {


The actual fixed differed from the expected fixed.
The actual fixed differed from the expected fixed.
Actual fixed saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/parser/issues/issue-105366/issue-105366.fixed
To only update this specific test, also pass `--test-args parser/issues/issue-105366.rs`

error: 2 errors occurred comparing output.
status: exit status: 1
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/parser/issues/issue-105366.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/parser/issues/issue-105366" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/parser/issues/issue-105366/auxiliary"
--- stderr -------------------------------
error: missing parameters for function definition
##[error]  --> /checkout/tests/ui/parser/issues/issue-105366.rs:5:13
   |
   |
LL | fn From<i32> for Foo {
   |             ^ help: add a parameter list

error: expected one of `->`, `where`, or `{`, found keyword `for`
   |
LL | fn From<i32> for Foo {
LL | fn From<i32> for Foo {
   |              ^^^ expected one of `->`, `where`, or `{`
error: aborting due to 2 previous errors
------------------------------------------



---- [ui] tests/ui/parser/removed-syntax-fn-sigil.rs stdout ----
diff of stderr:

- error: expected `(`, found `~`
2   --> $DIR/removed-syntax-fn-sigil.rs:2:14
3    |
3    |
4 LL |     let x: fn~() = || ();
-    |         -    ^ expected `(`
-    |         |
-    |         while parsing the type for `x`
+    |              ^ help: add a parameter list
+    |              ^ help: add a parameter list
8 
- error: aborting due to previous error
+ error: expected one of `->`, `;`, or `=`, found `~`
+    |
+    |
+ LL |     let x: fn~() = || ();
+    |              ^ expected one of `->`, `;`, or `=`
+ error: aborting due to 2 previous errors
10 
11 

---
--- stderr -------------------------------
error: missing parameters for function definition
##[error]  --> /checkout/tests/ui/parser/removed-syntax-fn-sigil.rs:2:14
   |
LL |     let x: fn~() = || (); //~ ERROR expected `(`, found `~`


error: expected one of `->`, `;`, or `=`, found `~`
   |
   |
LL |     let x: fn~() = || (); //~ ERROR expected `(`, found `~`
   |              ^ expected one of `->`, `;`, or `=`
error: aborting due to 2 previous errors
------------------------------------------


```

</details>

clubby789 commented:
Added some special cases for things that are handled specifically elsewhere. I also removed the failing test from `recovered-block.rs` as it's no longer testing what it was, and the old result didn't seem very meaningful

clubby789 commented on compiler/rustc_parse/src/parser/item.rs:
```diff
@@ -2505,7 +2505,29 @@ impl<'a> Parser<'a> {
             param
-        })?;
+        });
+        let mut params = match parsed {
```
I was originally doing it after to avoid trying to recover before we'd even encountered an error, and to avoid perf impact, but hopefully with inlining it should be okay

clubby789 commented:
@rustbot ready

petrochenkov commented on compiler/rustc_parse/src/parser/item.rs:
```diff
@@ -2492,6 +2492,19 @@ impl<'a> Parser<'a> {
+            // removed syntax, handled elsewhere
+            if self.token.kind != TokenKind::Tilde
+            // might be typo'd trait impl, handled elsewhere
+            && !self.token.is_keyword(kw::For)
```
Could you remove these special cases?
I don't think they are necessary here.

In `fn foo ~` -> `fn foo() ~` and `fn foo for` -> `fn foo() for` the unexpected tokens don't have any special meaning and can be reported equally well after inserting the `()`.

clubby789 replied:
There are some specialized diagnostics for `fn Trait for Type` (`impl` -> `fn`) typo and `fn~()` (removed syntax) which don't get reported if we recover and raise the diagnostic here

petrochenkov replied:
Ok, I see the `for` special case in item.rs, it's silly, but let's keep it.

I don't see any special logic for `~` though and I don't remember such syntax, which means it was removed in 2014 latest, so if any such special treatment exists it is no longer necessary.

petrochenkov commented:
Thanks!
@bors r+

bors commented:
:pushpin: Commit ca1bcb64667410c7f59ec5745e012601eea3d65f has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov ca1bcb64667410c7f59ec5745e012601eea3d65f -->
<!-- homu: {"type":"Approved","sha":"ca1bcb64667410c7f59ec5745e012601eea3d65f","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
